src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/GeneratedCodeContainer.cs

```diff
@@ -39,34 +95,59 @@ public TService GetService<TService>()
             return default(TService);
         }
 
-        public void SetOutput(SourceText source, RazorCodeDocument codeDocument)
+        public void SetOutput(DocumentSnapshot document)
         {
-            Source = source;
-            Output = codeDocument.GetCSharpDocument();
+            lock (_setOutputLock)
+            {
+                if (!document.TryGetTextVersion(out var version))
+                {
+                    return;
+                }
 
-            _textContainer.SetText(SourceText.From(Output.GeneratedCode));
+                var newerVersion = SourceVersion.GetNewerVersion(version);
+                if (newerVersion == SourceVersion)
+                {
+                    // Version from the provided document is the same or older as the current SourceVersion.
+                    return;
+                }
+
+                if (!document.TryGetText(out var source) ||
+                    !document.TryGetGeneratedOutput(out var codeDocument))
+                {
+                    return;
+                }
```

Comment:
Is this right? I thought this method was called before the output was available outside of the facade

Reply:
Referring to `TryGetGeneratedOutput`

Reply:
Ya I had to work around that restriction in a hackyyyy way: https://github.com/aspnet/Razor/pull/2572/files/8d9bf2e3cd4f4b4fe909e231c0bb130b586efbd2#diff-277383bfb275d15d91a911ee7b04ee3bR53

Reply:
this seems wrong then.

Reply:
Open to other suggestions, the only other one I thought of is having the `SetOutput` signature look like:

`SetOutput(RazorCodeDocument output, DefaultDocumentSnapshot document)`. Didn't like that though because it indicated that the code document didn't have to be from the passed in document. In practice of course that wouldn't be a problem; was just a code cleanliness concern.

Reply:
No, this is correct. You are currently modifying the state of the document snapshot.